File: app/src/main/java/de/westnordost/streetcomplete/view/ImageSelectAdapter.java
package de.westnordost.streetcomplete.view;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import de.westnordost.streetcomplete.R;

/** Select a number of items from a list of items */
public class ImageSelectAdapter extends RecyclerView.Adapter<ItemViewHolder>
{
	private ArrayList<Item> items = new ArrayList<>();
	private Set<Integer> selectedIndices;
	private int maxSelectableIndices;
	private int cellLayoutId = R.layout.cell_labeled_image_select;

	public interface OnItemSelectionListener
	{
		void onIndexSelected(int index);
		void onIndexDeselected(int index);
	}
	private final List<ImageSelectAdapter.OnItemSelectionListener> listeners = new ArrayList<>();

	public ImageSelectAdapter()
	{
		this(-1);
	}

	public ImageSelectAdapter(int maxSelectableIndices)
	{
		selectedIndices = new HashSet<>();
		this.maxSelectableIndices = maxSelectableIndices;
	}

	public void addOnItemSelectionListener(ImageSelectAdapter.OnItemSelectionListener listener)
	{
		listeners.add(listener);
	}

	public void setCellLayout(int cellLayoutId)
	{
		this.cellLayoutId = cellLayoutId;
	}

	public ArrayList<Integer> getSelectedIndices()
	{
		return new ArrayList<>(selectedIndices);
	}

	public List<Item> getSelectedItems()
	{
		List<Item> result = new ArrayList<>();
		for (Integer index : selectedIndices)
		{
			result.add(items.get(index));
		}
		return result;
	}

	public void select(List<Integer> indices)
	{
		for(Integer index : indices)
		{
			select(index);
		}
	}

	public void setItems(List<Item> items)
	{
		this.items = new ArrayList<>(items);
		notifyDataSetChanged();
	}

	@NonNull @Override
	public ItemViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType)
	{
		View view = LayoutInflater.from(parent.getContext()).inflate(cellLayoutId, parent, false);
		ItemViewHolder holder = new ItemViewHolder(view);
		holder.setOnClickListener(this::toggle);
		return holder;
	}

	public boolean isSelected(int index) { return selectedIndices.contains(index); }

	public void select(int index)
	{
		checkIndexRange(index);
		// special case: toggle-behavior if only one index can be selected
		if(maxSelectableIndices == 1 && selectedIndices.size() == 1)
		{
			deselect(selectedIndices.iterator().next());
		}
		else if(maxSelectableIndices > -1 && maxSelectableIndices <= selectedIndices.size())
		{
			return;
		}

		if(!selectedIndices.add(index)) return;

		notifyItemChanged(index);
		for (OnItemSelectionListener listener : listeners)
		{
			listener.onIndexSelected(index);
		}
	}

	public void deselect(int index)
	{
		checkIndexRange(index);
		if(!selectedIndices.remove(index)) return;

		notifyItemChanged(index);
		for (OnItemSelectionListener listener : listeners)
		{
			listener.onIndexDeselected(index);
		}
	}

	public void toggle(int index)
	{
		checkIndexRange(index);
		if(!isSelected(index))
		{
			select(index);
		}
		else
		{
			deselect(index);
		}
	}

	private void checkIndexRange(int index)
	{
		if(index < 0 || index >= items.size())
			throw new ArrayIndexOutOfBoundsException(index);
	}

	@Override public void onBindViewHolder(@NonNull ItemViewHolder holder, int position)
	{
		holder.bind(items.get(position));
		holder.setSelected(isSelected(position));
	}

	@Override public int getItemCount() { return items.size(); }
}

File: app/src/main/java/de/westnordost/streetcomplete/view/ImageSelectAdapter.kt
package de.westnordost.streetcomplete.view

import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup

import de.westnordost.streetcomplete.R
import java.util.concurrent.CopyOnWriteArrayList

/** Select a number of items from a list of items  */
class ImageSelectAdapter<T>(private val maxSelectableIndices: Int = -1) :
    RecyclerView.Adapter<ItemViewHolder>() {

    var items = listOf<Item<T>>()
    set(value) {
        field = value
        notifyDataSetChanged()
    }

    private val _selectedIndices = mutableSetOf<Int>()
    val selectedIndices get() = _selectedIndices.toList()

    var cellLayoutId = R.layout.cell_labeled_image_select

    val listeners: MutableList<OnItemSelectionListener> = CopyOnWriteArrayList()

    val selectedItems get() = _selectedIndices.map { i -> items[i].value!! }

    interface OnItemSelectionListener {
        fun onIndexSelected(index: Int)
        fun onIndexDeselected(index: Int)
    }

    fun indexOf(item: T): Int = items.indexOfFirst { it.value == item }

    fun select(indices: List<Int>) {
        for (index in indices) {
            select(index)
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ItemViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(cellLayoutId, parent, false)
        val holder = ItemViewHolder(view)
        holder.onClickListener = ::toggle
        return holder
    }

    fun isSelected(index: Int) = _selectedIndices.contains(index)

    fun select(index: Int) {
        checkIndexRange(index)
        // special case: toggle-behavior if only one index can be selected
        if (maxSelectableIndices == 1 && _selectedIndices.size == 1) {
            deselect(_selectedIndices.first())
        } else if (maxSelectableIndices > -1 && maxSelectableIndices <= _selectedIndices.size) {
            return
        }

        if (!_selectedIndices.add(index)) return

        notifyItemChanged(index)
        for (listener in listeners) {
            listener.onIndexSelected(index)
        }
    }

    fun deselect(index: Int) {
        checkIndexRange(index)
        if (!_selectedIndices.remove(index)) return

        notifyItemChanged(index)
        for (listener in listeners) {
            listener.onIndexDeselected(index)
        }
    }

    fun toggle(index: Int) {
        checkIndexRange(index)
        if (!isSelected(index)) {
            select(index)
        } else {
            deselect(index)
        }
    }

    private fun checkIndexRange(index: Int) {
        if (index < 0 || index >= items.size)
            throw ArrayIndexOutOfBoundsException(index)
    }

    override fun onBindViewHolder(holder: ItemViewHolder, position: Int) {
        holder.bind(items[position])
        holder.isSelected = isSelected(position)
    }

    override fun getItemCount() = items.size
}
